src/Parsers/ParserAlterQuery.cpp

```diff
@@ -870,6 +871,14 @@ bool ParserAlterCommand::parseImpl(Pos & pos, ASTPtr & node, Expected & expected
                     return false;
                 command->type = ASTAlterCommand::MODIFY_SQL_SECURITY;
             }
+            else if (s_modify_definer.ignore(pos, expected))
+            {
+                /// Same hack here
+                pos -= 1;
```

Comment:
Why do we need this hack here?

Reply:
`s_modify_definer` moves the current position to

```
ALTER TABLE MODIFY DEFINER = pufit
                          /\                                     
```
but the `ParserSQLSecurity` takes AST `DEFINER = ... [SQL SECURITY ...]` or `SQL SECURITY ... [DEFINER = ...]`
So, in order to reuse this parser, we need to move the current position one step back.

```
ALTER TABLE MODIFY DEFINER = pufit
                 /\                                     
```

The crash in the linked issue happened because `ASTSQLSecurity::formatImpl` returns a canonical form `DEFINER = ... SQL SECURITY DEFINER`, so the `ALTER` query `ALTER TABLE ... MODIFY SQL SECURITY DEFINER = pufit` changes after format into `ALTER TABLE ... MODIFY DEFINER = pufit SQL SECURITY DEFINER` and can't be parsed again.

Reply:
It's a bit too hacky, it seems better to modify `ParserSQLSecurity` and add there a parameter to its constructor which would allow to disable parsing either the `DEFINER` or `SQL SECURITY` keyword.